Add `Arc::into_unique` for converting back to `UniqueArc`.

tobz opened the pull request:
This is a spiritual equivalent to `std::sync::Arc::into_inner` where we can let shared `Arc` instances be atomically dropped or converted back to an owning `UniqueArc`.

## Notes

This is a reworking of #103, where we now use `AcqRel` ordering (instead of `Release`) when decrementing the strong count, in order to avoid a data race. We've also added an additional unit test that is able to exercise the data race _if_ the ordering was `Release` instead.

Admittedly, this unit test might not even properly detect the data race depending on nightly version and the instruction ordering in the compiled test binary... but it was much simpler to add than a full `loom` test, and hopefully can serve as some additional incremental coverage in the meantime?

tobz commented:
cc @steffahn

tobz commented:
Tests are now fixed after a little bit of wrangling.

Manishearth commented:
Probably should have a comment about the necessity of AcqRel (and the other options on the table).

I'll defer merging this until @steffahn has a look.

steffahn commented:
As noted in https://github.com/Manishearth/triomphe/pull/103#issuecomment-2562012106 `AcqRel` in this place is definitely sufficient; as would be `Release` paired with the subsequent `.load(Acquire)` like `Drop` does, and the `store` back to `1` doesn't need any ordering at all (not even atomic operations, necessarily).

So the implementation is fine this way; however as long as the comment explaining this code just states a simple
> ```Perform the same logic as `drop_inner` ```

I would expect that the could should do **exactly** as `drop_inner` does; i.e. the two-step `Release`-then-`Acquire` procedure. If you'd like to use `AcqRel`, there should be a short explanation that this has only stronger synchronization guarantees than the 2 steps in `drop_inner`. Either way, some comment should also explicitly point out the importance of some `Acquire` ordering being involved; in `drop_inner` that's a 20-line comment. Here, you could probably make it relatively short; e.g. referring the reader to the existing in-code documentation of `drop_inner` for more context, explaining that “deletion of the data” there, and “mutable access to the data” here is analogous; and perhaps even a mention of `https://github.com/Manishearth/triomphe/issues/106` could also be useful.

---

As for test case, I'd appreciate if something like
```rust
    let a = Arc::new(0);
    let b = a.clone();
    let t1 = std::thread::spawn(move || {
        let _value = *b;
    });
    let t2 = std::thread::spawn(move || {
        std::thread::sleep(std::time::Duration::from_millis(100));
        if let Some(mut u) = Arc::into_unique(a) {
            *u += 1
        }
    });
    t1.join().unwrap();
    t2.join().unwrap();
```
was also added (in addition to the existing one without any sleep) as that setup should more reliably catch the particular previous issue of insufficient synchronization, without relying too much on “random” execution order decisions of miri.

The existing sleep-less, join-less variant is useful, too, because it *can* have more different orders of execution. So assuming e.g. a setting where we (or someone else) would be running multiple rounds of `miri` tests with different seeds for randomness each time, then such a test can cover more ground in principle.

tobz commented:
@Manishearth @steffahn I've updated the inline comments around `fetch_sub(AcqRel)` and their importance, as well as added the additional unit test layout that @steffahn recommended.

Let me know if this addresses your concerns.

steffahn commented:
Looks good to me.

tobz commented:
@Manishearth Just checking back in: given @steffahn's review, would you consider this ready to merge?

tobz commented:
Gentle bump. :)

ambaxter commented:
bump, as I am also interested in this :)

Manishearth commented:
Sorry, I've been really busy.

Manishearth commented:
Published!

tobz commented:
Absolutely no problem: life happens. :)

Thanks for the approval and release!
